### ISSUE TITLE
Idea to treat blank-line-delimited lists as separate ones, instead of joining

### ISSUE THREAD

**Issue by john-cj** [asciidoctor/asciidoctor#3042]
To have multiple paragraphs inside one list item we use `+`. And to have two consequent lists we use `//` workaround. We are forced to use `//` to avoid lists to be joined.

```
Markup  |   Output              Markup  |   Output
        |                               |
. Foo   |   1. Foo              . Foo   |   1. Foo
. Bar   |   2. Bar              . Bar   |   2. Bar
+       |      Baz              +       |      Baz
Baz     |                       Baz     |   3. Xxx
        |   1. Xxx                      |   4. Yyy
//      |   2. Yyy              . Xxx   |
        |                       . Yyy   |
. Xxx   |                               |
. Yyy   |                               |
```

The idea is simple: treat blank-line-delimited lists as two separate ones, instead of joining them.

**Comment by elextr** [comment]
This would mean that all list items cannot be spaced or they won't be part of the same list.  That makes it hard to read them and breaks a large number of existing documents.

In general two consecutive lists without any intervening content is likely to be the least common situation, in almost all situations lists are separated by other content such as a paragraph introducing readers to the list.

Preventing a common case (list items separated by blank lines for readability) in favour of an uncommon case (consecutive lists) is bad design.

**Comment by john-cj** [comment]
For example, I never use spaced lists, because I like to have consistent look across all lists: large ones and small ones. And spacing looks very weird in small list.

Also, imagine the following situation: I have 10 lists in document, 8 very simple and two complex. Complex lists are spaced, simple ones aren't. Then, I need to add these simple ones to complex ones, to have only 2 lists in document. In this case, it will be necessary to remove empty lines from these simple lists, which are now part of the larger ones. This is why I always use non-spaced, as more reliable.

In other words, I understand your point, but I'm not sure spaced lists is a good thing.

**Comment by mojavelinux** [comment]
I've thought about this problem many times. I'd love to have an elegant way to solve it. I have yet to come up with a solution that doesn't have drawbacks. Here's my current proposal: https://github.com/asciidoctor/asciidoctor/issues/2639

As alluded to in the report, the most concise way to force adjacent lists apart is to start the list with either an empty block attribute list or a line comment. For example:

```
* first list
* first list
* first list

[]
* second list
* second list
* second list
```

(I like the empty block attribute line because it leaves room to grow into a non-empty block attribute line).

Without that indicator (or the double blank line I floated in my proposal), it's very difficult to work out the intent. When is it just a blank line for readability and when is it the start of a new list?

The main challenge with any change to how lists are interpreted is the risk of breaking scores of documents.

> john-cj wrote: \
> I never use spaced lists

You are but one user. I can assure you there are tons of documents that rely on this behavior, and plenty of users who prefer the spaced syntax.

> Complex lists are spaced, simple lists aren't.

That's my personal style.

> elextr wrote: \
> Preventing a common case (list items separated by blank lines for readability) in favour of an uncommon case (consecutive lists) is bad design.

I don't think we have to go so far as to say it's bad design. I understand the desire to have the parser understand when you want separate lists. But we are working within a set of constraints that we must honor. We don't have the luxury of inventing a new set of rules that don't honor existing rules (that's the part that would be bad).

Thus, we have to find a middle-ground that makes the new possible without completely disrupting the old. That's something I believe two blank lines proposal can achieve.

**Comment by elextr** [comment]
> I don't think we have to go so far as to say it's bad design.

A common tenet of good design for any language is that the common usage should be simple, concise and safe and the uncommon usage can be more verbose or complex.  The original proposal actually made the common case impossible, to allow the uncommon case to be concise, so I _would_ class that as bad design.

That does not mean that the uncommon case should be impossible, and there are several current solutions for it, eg a comment.

But so far I agree with @mojavelinux that none of the solutions are convincing:

1. using a comment isn't really a good solution, really comments should have no effect on the document parsing, it should just be like they were not there.

2. double line spacing, whitespace being significant is a risky design choice, its very easy to miss its significance, but it is used for paragraph separation so that boat has sailed :).  Enforcing "one blank line for paragraph breaking, two blank lines for list breaking" is possible, again its not a good choice from the point of view of being obvious and would be different to multiple blank lines anywhere else, eg I like to leave several blank lines before high level sections but they have no meaning.

3. empty attribute list, this is not so good because a non-empty attribute list won't separate the lists, because if it did there would be no way of adding attributes to list items inside the list

In general the Asciidoc list syntax and semantics have several hairy edge cases that need addressing, mostly due to there being no list start or list end indication besides a change of entity (eg paragraph, list item means start of list, list item, paragraph means end of list)  This is consistent with the rest of Asciidoc syntax and semantics for streamed entities of course, but other streamed entities (eg sections) cannot break streams.  Thats why the list breaking doesn't seem to fit in, it doesn't, it is different.

So lists being different, the thing we should consider is something different, such as an optional explicit list start marker instead of co-opting other syntax as a special case.  That also allows attributes for the list, and the first list item to be clearly separated if needed.

Some suggestions:

1. an empty list item the same as that used for the list

2. a separate syntax common for all lists, but we are running out of Ascii characters.

Note I said optional, lists without an explicit marker can behave as they do now. So its backward compatible.

More thought is needed, and its good that @john-cj prompted it.

**Comment by mojavelinux** [comment]
> so I _would_ class that as bad design.

What I'm trying to say nicely is that this is not welcoming language for this forum. Perhaps the idea does not take into account all usages, but saying it's bad just shuts down debate.

**Comment by mojavelinux** [comment]
> empty attribute list, this is not so good because a non-empty attribute list won't separate the lists, because if it did there would be no way of adding attributes to list items inside the list

This already works and has always worked. A block attribute list is a block breaking line.

**Comment by mojavelinux** [comment]
The fact that lists don't have explicit boundary markers (aka fences) is consistent with all other lightweight markup languages and we don't want to violate that (or else it would make AsciiDoc really unappealing). So the focus should be on this question:

"How can we tell when a new list should be started?" or the corollary, "How can we tell when a list is supposed to end?"

Building on something lex suggested, I think we're looking for a prelude or coda marker. Arguably, we already have one for the start of a list with the empty block attribute list. We could just document it and make it a bit more prominent.

```
* first list

[]
* second list
```

But I'd also like to see it possible to use a coda mark, since often it's the end of the list you want to mark.

```
* first list
<- something here

* second list
```

We could consider the inverse of a line comment:

```
* first list
\\

* second list
```

But we need to move on from the idea that it can be a single line. That's just not enough to indicate the list should end. Doing so would introduce one of the main gripes of Markdown, which is that lists are really hard to keep together. (We just happen to have the opposite problem).

**Comment by elextr** [comment]
> What I'm trying to say nicely is that this is not welcoming language for this forum. Perhaps the idea does not take into account all usages, but saying it's bad just shuts down debate.

I guess we are suffering from the English speakers problem, assuming that our cultural backgrounds are the same just because we both speak English, and applying our local nuances to things that are said.  Here calling something bad is not an end of the conversation, as should be clear from the size of the rest of the post.  But since this is your forum, its fine for you to say you don't want it to be said that way because you see it as a "bad thing" :).  What word would you use other than "bad"?

> > empty attribute list, this is not so good because a non-empty attribute list won't separate the lists, because if it did there would be no way of adding attributes to list items inside the list

> This already works and has always worked. A block attribute list is a block breaking line.

I said that badly, there is a situation where you might want to apply attributes to list items without breaking lists that isn't currently accommodated by Asciidoc syntax/semantics.  I remember some discussion about putting the attribute list after the list item markup, but I couldn't find any reference to that in the manual (but both list and attribute sections are big and I might have missed it).

**Comment by elextr** [comment]
> But we need to move on from the idea that it can be a single blank line. That's just not enough to indicate the list should end. Doing so would introduce one of the main gripes of Markdown, which is that lists are really hard to keep together. (We just happen to have the opposite problem).

+millions for not following markdown here, and have you seen the size of the list section of the commonmark spec :grin:

I'd like to think on it some more, as there are current workarounds its not urgent that something be implemented quickly.  For the purposes of separating lists either list start or list end would be ok, I suggested list start because it might also help address the attribute issue.

**Comment by elextr** [comment]
@john-cj interesting suggestion (and I like your analogy `list1 != list2` :).  

The downside is that it means there is a second set of all of the list tokens that start with `!`.   Thats not a killer, but its a consideration as there are a number of list types.

When lists are not autonumbered I think the `!` would be easy to miss on a quick reading.

```
11!. a new list item 11

not a continuation!::: of the previous list, but a new list
```

And that last is something none of us have addressed to date is how this feature would fit with labelled lists.

**Comment by mojavelinux** [comment]
For me, "bad design" invokes very negative feelings. What we have here are good intentions. We just need to figure out how to get from good intentions to good design ;)

**Comment by mojavelinux** [comment]
> From my point of view, the marker should be located on top, instead of on bottom, to be consistent with existing Asciidoctor rules.

Please feel free to make either case.

**Comment by john-cj** [comment]
@elextr 

> And that last is something none of us have addressed to date is how this feature would fit with labelled lists.

It seems labels are automatically delimit lists, so you don't need to have additional `!`:

```
. one
. two

.label
. one
. two
```

However, if we want to use additional `!` with label, then it will be:

```
// Exclamation mark isn't required in this example

. one
. two

.label
!. first item
. second
```

...and then we could remove the label, if we wish.

**Comment by elextr** [comment]
@john-cj oops, I was using the old Asciidoctor name for description list, sorry if you thought I was talking about titles.  See the second example I gave, the ! is pretty hidden.

```
not a continuation!::: of a previous list, but a new list
```

**Comment by elextr** [comment]
> This doesn't work for me in Asciidoctor (it works in Markdown only)

The manual says it should work (although the second is preferred).

One of my long term beefs with (AFAIK) all Asciidoc implementations is this insistence that lists be numbered sequentially.  I have worked in government and aerospace industries and in those arenas lists don't get renumbered even when items are removed, that way references from other communications continue to point to the same item.  This makes non-sequential numberings occur with a sufficient frequency that it would preclude using Asciidoc in those applications.  `</rant, sorry>`

But clearly I don't want to preclude manual numbered lists, something you could not have been expected to know, but as I said the manual says Asciidoctor should work for a different starting number.

Your section about definition lists isn't what I'm talking about. What I'm talking about has nothing to do with mixed list types, just the same question regarding breaking definition lists.  Hopefully a longer example will be clearer:

```
a definition:: a definition list item

another definition:: another definition list item

[]
another definition:: in a new list if the [] works? or will a // comment work?

definition of this:: some more of the new list

definition of that!:: this is a new definition list if you noticed the `!`, but it could also be part of the term being defined, I guess this is what inspired your very last line moving the ! to the start

definition of the other:: more of the second definition list
```

The syntax:

```
!text:: more text
```

is unlike any other elements in Asciidoc, because a single markup semantic is split across intervening text.  All other markups enclosing text have start and end semantics.  It makes it easier for writers if there are fewer variations in Asciidoc behaviour.

The more I think about it the more convinced I am that something on a different line should be used to signal list splitting.  Its unusual, so it should be clear to reduce erroneous reads by humans.

**Comment by john-cj** [comment]
@elextr Ah, I understand. You mean the issue of separating two definition lists one from another, right?

> The more I think about it the more convinced I am that something on a different line should be used to signal list splitting. Its unusual, so it should be clear to reduce erroneous reads by humans.

Same thoughts. Exclamation mark works good for delimiting normal lists, but it doesn't work good for definition lists, as was just discovered. So, it seems `[]` as Dan suggested is the best way.

**Comment by john-cj** [comment]
Another option could be a switch. Something like `:tightlists:`.
